### PR TITLE
feat: google authentication backend

### PR DESCRIPTION
**Pull request by Santosh7017**
### Related Issue 

ISSUE NUMBER #711 

#### Changes made 👷🏻‍♂️

This feature introduces the backend implementation of enhanced authentication functionality by integrating Google Account authentication. With this update, users will be able to securely authenticate and access the application using their Google Account credentials, providing an additional layer of authentication and simplifying the login process. This feature enhances the security and user experience of the application, ensuring reliable and convenient access for all users. The changes made in this pull request effectively enhance the overall authentication system, thereby strengthening the security and usability of the application.

## Necessary steps to setup .env

these are the values of n .env file
CLIENT_ID = "" // OAuth client ID
CLIENT_SECRET = "" // OAuth client secret
CALLBACK_URL = "" // callback URL like (http://localhost:5000/auth/google/callback) in place of localhost put the live server's link where you will receive callback
successURL = "" // this URL will redirect user to the page where we want send him after succesfull authentication for example Home page URL(https://milaan.vercel.app/)

### To create a Google OAuth client ID and client secret, you need to follow the steps below:

- Go to the Google Cloud Console: Visit the Google Cloud Console at https://console.cloud.google.com/ and sign in with your Google account.
- Create a new project: If you don't have a project yet, create a new project by clicking on the project drop-down menu at the top of the page and selecting "New Project". Give your project a name and click "Create".
- Enable the necessary APIs: In the Cloud Console, go to the "APIs & Services" section. Click on "Library" in the left-hand menu. Search for and enable the following APIs:
- Google OAuth 2.0 API
- Set up OAuth consent screen: In the Cloud Console, navigate to the "APIs & Services" section and click on "OAuth consent screen" in the left-hand menu. Choose an appropriate user type and fill in the required information, such as the application name and authorized domains. 
- Create OAuth client ID: In the Cloud Console, go to the "APIs & Services" section and click on "Credentials" in the left-hand menu. Click on the "Create Credentials" button and select "OAuth client ID" from the drop-down menu.

- Configure the OAuth client ID: Choose the application type, such as "Web application" or "Android/iOS app". Enter the necessary details, including authorized JavaScript origins or redirect URIs, depending on your application type. Save the configuration.

- Obtain the client ID and client secret: Once you've created the OAuth client ID, the client ID and client secret will be displayed on the "Credentials" page. Make sure to copy and securely store these values as they are required for authenticating your application

#### to know more you can visit - https://support.google.com/cloud/answer/6158849?hl=en

### CODE REVIEW

File: server/index.js [hunk 1]
```diff
@@ -11,12 +11,36 @@ let port = process.env.PORT || 5000;
...
+app.use(passport.initialize());
+app.use(passport.session());
+
 app.get("/", (req, res) => {
   res.send("HELLO FROM HOME");
 });
 
 //* user routes
-app.use("/user", require("./routes/User"));
+app.use("/user", require("./routes/user"));
```
Comment (tamalCodes) [inline server/index.js:43]: Please revert this back to `./routes/User` as we have the file name as `User.js` having small letter can sometimes give rise to bugs.

File: server/package.json [hunk 2]
```diff
@@ -12,13 +12,19 @@
...
     "dotenv": "^15.0.0",
     "express": "^4.17.2",
+    "express-session": "^1.17.3",
     "jsonwebtoken": "^9.0.0",
     "mongoose": "^6.1.10",
     "nodemon": "^2.0.20",
+    "passport": "^0.6.0",
+    "passport-google-oauth": "^2.0.0",
+    "passport-google-oauth2": "^0.2.0",
+    "passport-google-oauth20": "^2.0.0",
```
[the remark this answers is not in the document]
Comment (tamalCodes) [reply server/package.json:25]: Great !

**Review by tamalCodes — CHANGES_REQUESTED** [review]
The changes look fine to me, can you perhaps record a short screencast (tools like Loom, OBS can be of help) and just show that the backend is working as intended ?

**Review by tamalCodes — APPROVED** [review]
**Amazing work,**
This looks good to me, will be merged soon 🎉.

**Comment by Santosh7017** [comment]
It is pleasing to receive this information. what next to do, Should I proceed with closing the current issue and opening a new one for the frontend?

**Comment by tamalCodes** [comment]
> It is pleasing to receive this information. what next to do, Should I proceed with closing the current issue and opening a new one for the frontend?

Yaa sure go ahead !

**Pull request merged**